File: LL.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedLists:
    def __init__(self, value=0):
        self.length = 0
        if value is None:
            return
        newNode = Node(value=value)
        self.head = newNode
        self.tail = newNode
        self.length += 1
    
    def print_list(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next
    
    def append(self, value):
        newNode = Node(value=value)
        if not self.head:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return True
    
    def pop(self):
        if self.length == 0: return
        temp = self.head
        prev = self.head
        while (temp.next):
              prev = temp
              temp = temp.next
        prev.next = None
        self.tail = prev
        self.length -= 1
        if self.length == 0:
            self.head = None
            self.tail = None
            return temp.value
    
    def prepend(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        if self.length == 0: self.tail = new_node 
        self.length += 1
        
        return True
    
    def pop_first(self):
        if self.length == 0: return None
        temp = self.head
        self.head = self.head.next
        temp.next == None
        self.length -= 1
        if self.length == 0:
            self.tail = None
        return temp
    
    def get(self, index):
        if index < 0 or index > self.length: return None
        temp = self.head
        for _ in range(index): temp = temp.next
        return temp
    def set_value(self, index, value):
        temp = self.get(index=index)
        if temp: 
            temp.value = value
            return True
        return False
    def insert(self, index, value):
        if index < 0 or index > self.length: return False
        if index == 0:
            return self.prepend(value=value)
        if index == self.length:
            return self.append(value=value)
        temp = self.get(index)
        new_node = Node(value=value)
        new_node.next = temp.next
        temp.next = new_node
        self.length += 1
        return True
    def remove(self, index):
        if index < 0 or index >= self.length: return False
        if index == 0: return self.pop_first()
        if index == self.length - 1: return self.pop()
        temp = self.head
        prev = self.head
        for _ in range(index):
            prev = temp
            temp = temp.next
        prev.next = temp.next
        return temp
    def getLength(self):
        length = 0
        temp = self.head
        while temp:
            length += 1
            temp = temp.next
        return length
    
    def find_middle_node(self):
        length = self.getLength()
        if length == 0: return None
        temp = self.head
        if length%2 == 0:
            loop = (length/2)
            for _ in range(int(loop)):
                temp = temp.next
        else:
            loop = ((length + 1)/2)
            for _ in range((int((loop)))- 1):
                temp = temp.next
        return temp.value
    
    def reverse(self):
        temp = self.head
        self.head = self.tail
        self.tail = temp
        after = temp.next
        before = None
        for _ in range(self.length):
            after = temp.next
            temp.next = before
            before = temp
            temp = after
        return True

myLL = LinkedLists(1)
myLL.append(2)
myLL.append(3)
myLL.append(4)
myLL.print_list()
print(f"Middle: {myLL.find_middle_node()}")
print(f"Length: {myLL.length}")
print(f"Head: {myLL.head.value if myLL.head is not None else "None"}")
print(f"Tail: {myLL.tail.value if myLL.tail is not None else "None"}")

File: LL.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor(value = 0) {
        this.length = 0;
        if (value !== null) {
            const newNode = new Node(value);
            this.head = newNode;
            this.tail = newNode;
            this.length += 1;
        } else {
            this.head = null;
            this.tail = null;
        }
    }

    printList() {
        let temp = this.head;
        while (temp) {
            console.log(temp.value);
            temp = temp.next;
        }
    }

    append(value) {
        const newNode = new Node(value);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length += 1;
        return true;
    }

    pop() {
        if (this.length === 0) return null;
        let temp = this.head;
        let prev = this.head;
        while (temp.next) {
            prev = temp;
            temp = temp.next;
        }
        prev.next = null;
        this.tail = prev;
        this.length -= 1;
        if (this.length === 0) {
            this.head = null;
            this.tail = null;
            return temp.value;
        }
        return temp.value;
    }

    prepend(value) {
        const newNode = new Node(value);
        newNode.next = this.head;
        this.head = newNode;
        if (this.length === 0) this.tail = newNode;
        this.length += 1;
        return true;
    }

    popFirst() {
        if (this.length === 0) return null;
        const temp = this.head;
        this.head = this.head.next;
        temp.next = null;
        this.length -= 1;
        if (this.length === 0) {
            this.tail = null;
        }
        return temp;
    }

    get(index) {
        if (index < 0 || index >= this.length) return null;
        let temp = this.head;
        for (let i = 0; i < index; i++) {
            temp = temp.next;
        }
        return temp;
    }

    setValue(index, value) {
        const temp = this.get(index);
        if (temp) {
            temp.value = value;
            return true;
        }
        return false;
    }

    insert(index, value) {
        if (index < 0 || index > this.length) return false;
        if (index === 0) return this.prepend(value);
        if (index === this.length) return this.append(value);
        const temp = this.get(index);
        const newNode = new Node(value);
        newNode.next = temp.next;
        temp.next = newNode;
        this.length += 1;
        return true;
    }

    remove(index) {
        if (index < 0 || index >= this.length) return false;
        if (index === 0) return this.popFirst();
        if (index === this.length - 1) return this.pop();
        let temp = this.head;
        let prev = this.head;
        for (let i = 0; i < index; i++) {
            prev = temp;
            temp = temp.next;
        }
        prev.next = temp.next;
        return temp;
    }

    getLength() {
        return this.length;
    }

    findMiddleNode() {
        const length = this.getLength();
        if (length === 0) return null;
        let temp = this.head;
        const loop = length % 2 === 0 ? length / 2 : (length + 1) / 2;
        for (let i = 0; i < Math.floor(loop); i++) {
            temp = temp.next;
        }
        return temp.value;
    }

    reverse() {
        let temp = this.head;
        this.head = this.tail;
        this.tail = temp;
        let after = temp.next;
        let before = null;
        for (let i = 0; i < this.length; i++) {
            after = temp.next;
            temp.next = before;
            before = temp;
            temp = after;
        }
        return true;
    }
}

// Example usage
const myLL = new LinkedList(1);
myLL.append(2);
myLL.append(3);
myLL.append(4);
myLL.printList();
console.log(`Middle: ${myLL.findMiddleNode()}`);
console.log(`Length: ${myLL.length}`);
console.log(`Head: ${myLL.head ? myLL.head.value : "None"}`);
console.log(`Tail: ${myLL.tail ? myLL.tail.value : "None"}`);
